kernel/comps/keyboard/src/i8042_chip/controller.rs

```diff
@@ -0,0 +1,282 @@
+        controller.wait_and_send_command(Command::DisableFirstPort)?;
+        controller.wait_and_send_command(Command::DisableSecondPort)?;
+
+        // Flush the the output buffer by reading from the data port and discarding the data.
+        while controller.receive_data().is_some() {}
+
+        // Set the controller configuration byte.
+        controller.update_configuration(|config| {
+            config.remove(
+                Configuration::FIRST_PORT_INTERRUPT_ENABLED
+                    | Configuration::FIRST_PORT_TRANSLATION_ENABLED
+                    | Configuration::SECOND_PORT_INTERRUPT_ENABLED,
+            );
+        })?;
+
+        let config = controller.read_configuration()?;
+        // Perform controller self-test.
+        controller.wait_and_send_command(Command::TestController)?;
+        let result = controller.wait_and_recv_data()?;
+        if result != 0x55 {
+            // Any value other than 0x55 indicates a self-test fail.
+            return Err(I8042ControllerError::ControllerTestFailed);
+        }
+        // The self-test may reset the controller. Restore the original configuration.
+        controller.write_configuration(config)?;
+
+        // Perform interface tests to the first PS/2 port.
+        controller.wait_and_send_command(Command::TestFirstPort)?;
+        let result = controller.wait_and_recv_data()?;
+        if result != 0x00 {
+            return Err(I8042ControllerError::FirstPortTestFailed);
+        }
+
+        // Perform interface tests to the second PS/2 port (if it exists).
+        controller.wait_and_send_command(Command::EnableSecondPort)?;
+        let has_second_port = !controller
+            .read_configuration()?
+            .contains(Configuration::SECOND_PORT_CLOCK_DISABLED);
+        if has_second_port {
+            controller.wait_and_send_command(Command::DisableSecondPort)?;
+            controller.wait_and_send_command(Command::TestSecondPort)?;
+            let result = controller.wait_and_recv_data()?;
+            if result != 0x00 {
+                return Err(I8042ControllerError::SecondPortTestFailed);
+            }
+        }
+
+        // Enable the first PS/2 port.
+        controller.wait_and_send_command(Command::EnableFirstPort)?;
+        if super::keyboard::init(&mut controller).is_ok() {
+            controller.update_configuration(|config| {
+                config.insert(
+                    Configuration::FIRST_PORT_INTERRUPT_ENABLED
+                        | Configuration::FIRST_PORT_TRANSLATION_ENABLED,
+                )
+            })?;
+        } else {
+            controller.wait_and_send_command(Command::DisableFirstPort)?;
+        }
+
+        // TODO: Add a mouse driver and enable the second PS/2 port (if it exists).
+
+        Ok(controller)
+    }
+
+    pub(super) fn update_configuration<F>(&mut self, f: F) -> Result<(), I8042ControllerError>
```

Comment:
We may want to remove the `update_configuration` method since, after addressing the above comment, it seems not very useful.

I still don't know how you're ordering the methods, though. Do you plan on grouping methods with the same visibility together? Otherwise, I don't understand why you wouldn't put `read_configuration`/`write_configuration` here.

Personally, I still think that ordering the functions according to the calling stack would be better. In that case, `read_configuration` and `write_configuration` should be above `wait_and_send_command`.